meddlr/metrics/ssfd.py

```diff
@@ -0,0 +1,260 @@
+        layer_names: Sequence[str] = ["block4_relu2"],
+        channel_names: Sequence[str] = None,
+        reduction="none",
+        compute_on_step: bool = False,
+        dist_sync_on_step: bool = False,
+        process_group: bool = None,
+        dist_sync_fn: bool = None,
+    ):
+        """
+        Args:
+            mode (str): Determines how to interpret the channel dimension of the inputs. One of:
+                * ``'grayscale'``: Each channel corresponds to a distinct grayscale input image.
+                * ``'rgb'``: The 3 channel dimensions correspond to a single rgb image.
+                             Exception will be thrown if channel dimension != 3 or dtype is complex
+            layer_names (Sequence[str]):
+                A list of strings specifying the layers to extract features from. Any of:
+                ['block1_relu2', 'block2_relu2', 'block3_relu2', 'block4_relu2', 'block5_relu2']
+                SSFD from each layer will be summed if multiple layers are specified.
+        """
+
+        super().__init__(
+            channel_names=channel_names,
+            units="",
+            reduction=reduction,
+            compute_on_step=compute_on_step,
+            dist_sync_on_step=dist_sync_on_step,
+            process_group=process_group,
+            dist_sync_fn=dist_sync_fn,
+        )
+
+        valid_modes = ("grayscale", "rgb")
+        if mode not in valid_modes:
+            raise ValueError(f"Invalid `mode` ('{mode}'). Expected one of {valid_modes}.")
+
+        self.mode = mode
+        self.layer_names = layer_names
+
+        path_manager = env.get_path_manager()
+        file_path = path_manager.get_local_path(SSFD_HUGGINGFACE_URL, force=False)
+
+        self.net = SSFD_Encoder(in_channels=1, out_channels=320)
+        self.net.load_state_dict(torch.load(file_path))
+        self.net.eval()
+
+    def func(self, preds, targets) -> torch.Tensor:
+
+        if self.mode == "grayscale":
+            loss_shape = (targets.shape[0], targets.shape[1])
+
+        elif self.mode == "rgb":
+            if targets.shape[1] != 3:
+                raise ValueError(
+                    f"Channel dimension must have size 3 for rgb mode,\
+                    but got tensor of shape {targets.shape}."
+                )
+
+            is_complex = cplx.is_complex(targets) or cplx.is_complex_as_real(targets)
+            if is_complex:
+                raise TypeError(
+                    f"Data type must be real when mode is {self.mode},\
+                    but got data type {targets.dtype}"
+                )
+
+            loss_shape = (targets.shape[0], 1)
+
+        preds = self.preprocess_ssfd(preds)
+        targets = self.preprocess_ssfd(targets)
+
+        target_features = self.net(targets)
+        pred_features = self.net(preds)
+
+        loss = 0
+        for layer in self.layer_names:
+            loss += torch.mean(mse(target_features[layer], pred_features[layer]), dim=1)
```

Comment:
should dim be everything except the batch dimension? Is it possible to have features of shape BxCxHxW for example?